### ISSUE TITLE
feat: Fix ML Algorithm Implementations in Web Playground

### ISSUE THREAD

**Issue by BrunodsLilly** [BrunodsLilly/RustML#5]
## Overview

The ML Playground web interface (`web/src/components/ml_playground.rs`) has fully functional algorithm implementations that are currently commented out due to incorrect trait syntax. All algorithms (K-means, PCA, Logistic Regression, StandardScaler, MinMaxScaler) have working implementations in their respective crates with comprehensive test coverage. This issue tracks uncommenting and fixing the trait method calls to enable real ML computation in the browser.

## Problem Statement

**Current State:**
- Lines 350-374: All algorithms return placeholder "🚧 Full implementation coming soon!" messages
- Lines 376-517: Fully implemented algorithm runners are commented out
- Build succeeds but users cannot actually run ML algorithms on their CSV data

**Root Cause:**
Incorrect trait type parameter ordering and unnecessary explicit trait syntax. The commented code uses:
- `Clusterer<Matrix<f64>>` (missing T parameter) 
- `UnsupervisedModel<Matrix<f64>, f64>` (reversed parameter order)
- `Scaler<Matrix<f64>, f64>` (reversed parameter order)

**Correct Pattern:**
All traits follow `<T: Numeric, D: Data<T>>` ordering where T=f64, D=Matrix<f64>:
- `Clusterer<f64, Matrix<f64>>`
- `UnsupervisedModel<f64, Matrix<f64>>`
- `Scaler<f64, Matrix<f64>>`

**Impact:**
Users cannot test real ML algorithms, only see preview messages. This blocks the primary value proposition of the ML Playground.

## Proposed Solution

**Simplify to Method Call Syntax**

Instead of explicit trait syntax, use direct method calls and let Rust's type inference handle trait resolution:

\`\`\`rust
// ❌ Current (commented out, wrong syntax)
<KMeans as Clusterer<Matrix<f64>>>::fit(&mut kmeans, &dataset.features)

// ✅ Proposed (simple, correct)
kmeans.fit(&dataset.features)
\`\`\`

This works because:
1. KMeans implements \`Clusterer<f64, Matrix<f64>>\`
2. The method signature is unambiguous
3. Rust infers the trait implementation from the receiver type

### File to Modify
\`web/src/components/ml_playground.rs\`

#### 1. Uncomment Algorithm Runners (Lines 376-517)

Remove the comment markers \`/*\` and \`*/\` around existing implementations.

#### 2. Fix Each Algorithm's Trait Calls

**KMeans (lines 379-392):**
\`\`\`rust
// Before (line 365):
match <KMeans as Clusterer<Matrix<f64>>>::fit(&mut kmeans, &dataset.features) {

// After:
match kmeans.fit(&dataset.features) {
\`\`\`

Similarly for \`predict()\` call on line 367.

**PCA (lines 394-417):**
\`\`\`rust
// Before (line 398):
match <PCA as UnsupervisedModel<Matrix<f64>, f64>>::fit(&mut pca, &dataset.features) {

// After:
match pca.fit(&dataset.features) {
\`\`\`

Similarly for \`transform()\` call on line 400.

**Logistic Regression (lines 419-458):**
\`\`\`rust
// Before (lines 424-426):
let X_rows: Vec<Matrix<f64>> = (0..dataset.features.rows)
    .map(|i| dataset.features.row(i).unwrap())
    .collect();
match <LogisticRegression as SupervisedModel<f64, Matrix<f64>>>::fit(&mut model, &X_rows, &dataset.targets) {

// After:
match model.fit(&dataset.features, &dataset.targets) {
\`\`\`

Remove the X_rows conversion entirely - the trait accepts \`&Matrix<f64>\` directly.

Similarly for \`predict()\` call.

**StandardScaler (lines 460-473):**
\`\`\`rust
// Before (line 458):
match <StandardScaler as Scaler<Matrix<f64>, f64>>::fit(&mut scaler, &dataset.features) {

// After:
match scaler.fit(&dataset.features) {
\`\`\`

Similarly for \`transform()\` call.

**MinMaxScaler (lines 475-494):**
\`\`\`rust
// Before (line 478):
match <MinMaxScaler as Scaler<Matrix<f64>, f64>>::fit(&mut scaler, &dataset.features) {

// After:
match scaler.fit(&dataset.features) {
\`\`\`

Similarly for \`transform()\` call.

#### 3. Update run_algorithm() Dispatcher

Replace placeholder implementations (lines 350-374) with calls to the uncommented runners:

\`\`\`rust
fn run_algorithm(algorithm: Algorithm, dataset: &CsvDataset) -> String {
    match algorithm {
        Algorithm::KMeans => run_kmeans(dataset),
        Algorithm::PCA => run_pca(dataset),
        Algorithm::LogisticRegression => run_logistic_regression(dataset),
        Algorithm::StandardScaler => run_standard_scaler(dataset),
        Algorithm::MinMaxScaler => run_minmax_scaler(dataset),
    }
}
\`\`\`

### Functional Requirements

- [ ] All 5 algorithms execute when "▶ Run Algorithm" button is clicked
- [ ] K-means displays cluster assignments (e.g., "Cluster 0: 15 samples, Cluster 1: 12 samples, Cluster 2: 8 samples")
- [ ] PCA shows variance explained percentages and dimension reduction
- [ ] Logistic Regression calculates and displays accuracy percentage
- [ ] StandardScaler shows feature count and transformation confirmation
- [ ] MinMaxScaler shows feature count and scaling confirmation
- [ ] Error messages display for invalid data (empty CSV, incompatible shapes, etc.)

### Non-Functional Requirements

- [ ] Build succeeds: \`dx build --package web --platform web\`
- [ ] No compiler warnings related to the changes
- [ ] Algorithms complete within 2 seconds for datasets < 1000 samples
- [ ] UI remains responsive during computation (async spawn works correctly)

### Quality Gates

- [ ] Code compiles without errors or trait resolution issues
- [ ] Manual testing with sample CSV files confirms all algorithms work
- [ ] Result messages are clear, informative, and properly formatted
- [ ] No panic/unwrap in WASM code paths (all errors return Result)

### Step 1: Verify Test Coverage

Before uncommenting, confirm all algorithms work in their native crates:

\`\`\`bash
cargo test -p clustering      # 12 tests for KMeans
cargo test -p dimensionality_reduction  # 8 tests for PCA
cargo test -p preprocessing   # 11 tests for scalers
cargo test -p supervised      # 10 tests for LogisticRegression
\`\`\`

All should pass (they already do as of commit c859563).

### Step 2: Make Changes

Edit \`web/src/components/ml_playground.rs\`:

1. Remove comment block markers (lines 376, 517)
2. Simplify trait calls to method syntax (5 algorithms × 2 methods each = 10 changes)
3. Remove X_rows conversion in logistic regression
4. Update run_algorithm() dispatcher to call uncommented functions

### Step 3: Build and Test

\`\`\`bash
# Build for WASM
dx build --package web --platform web

# Serve locally
dx serve --package web

# Navigate to http://localhost:8080/playground
# Upload a CSV file (e.g., iris.csv, titanic.csv)
# Test each algorithm
\`\`\`

### Step 4: Create Test CSV Files (Optional)

**test_clustering.csv** (for K-means, PCA):
\`\`\`csv
feature1,feature2,feature3
1.2,2.3,0.5
1.5,2.1,0.6
5.0,8.2,3.1
4.8,8.5,2.9
9.1,1.2,5.4
8.9,1.5,5.2
\`\`\`

**test_classification.csv** (for Logistic Regression):
\`\`\`csv
feature1,feature2,class
1.0,2.0,0
2.0,3.0,0
6.0,7.0,1
7.0,8.0,1
\`\`\`

## Success Metrics

- **Accuracy**: All algorithms produce mathematically correct results (verified against test suites)
- **Performance**: Algorithms complete in < 2s for small datasets (< 1000 samples)
- **Usability**: Result messages are clear enough for non-ML users to understand
- **Reliability**: No crashes or silent failures in WASM environment

### Dependencies
- All ML algorithm crates already compiled and working ✅
- Trait implementations tested and validated ✅
- Dioxus async spawn pattern proven (optimizer demo) ✅

### Risks & Mitigation

| Risk | Impact | Mitigation |
|------|--------|-----------|
| Large datasets freeze UI | High | Add progress callback, limit to 10k samples |
| Algorithm convergence issues | Medium | Use tested default parameters from test suites |
| WASM memory limits | Medium | Bounded buffers already implemented |
| Confusing error messages | Low | Wrap errors with user-friendly context |

### Internal Code References

**Trait Definitions:**
- \`ml_traits/src/clustering.rs:7-26\` - Clusterer trait
- \`ml_traits/src/unsupervised.rs:16-30\` - UnsupervisedModel trait  
- \`ml_traits/src/preprocessing.rs:16-35\` - Scaler trait
- \`ml_traits/src/supervised.rs:25-40\` - SupervisedModel trait

**Working Implementations:**
- \`clustering/src/kmeans.rs:237-313\` - KMeans impl Clusterer
- \`dimensionality_reduction/src/pca.rs:157-280\` - PCA impl UnsupervisedModel
- \`preprocessing/src/scalers.rs:77-198\` - StandardScaler impl Scaler
- \`preprocessing/src/scalers.rs:273-397\` - MinMaxScaler impl Scaler
- \`supervised/src/logistic_regression.rs:276-342\` - LogisticRegression impl SupervisedModel

**Test Examples:**
- \`clustering/src/kmeans.rs:315-423\` - 12 comprehensive tests
- \`dimensionality_reduction/src/pca.rs:282-389\` - 8 validation tests
- \`preprocessing/src/scalers.rs:586-726\` - 11 scaler tests
- \`supervised/src/logistic_regression.rs:344-467\` - 10 classification tests

**Commented-Out Code:**
- \`web/src/components/ml_playground.rs:376-517\` - Full implementations ready to enable

### External Documentation

- **Rust Trait System**: [Trait Objects vs Generics](https://doc.rust-lang.org/book/ch17-02-trait-objects.html)
- **Type Inference**: [Rust RFC #213](https://github.com/rust-lang/rfcs/blob/master/text/0213-defaulted-type-params.md)
- **Dioxus Async**: [Spawning Futures](https://dioxuslabs.com/learn/0.6/reference/async)
- **WASM Best Practices**: Research compiled in \`docs/RUST_WASM_ML_BEST_PRACTICES.md\`

### Similar Work

- Optimizer Demo (\`web/src/components/optimizer_demo.rs\`) - Proven async pattern achieving 1000+ iter/sec
- Showcase (\`web/src/components/showcase.rs\`) - Matrix operations working in WASM

## Future Considerations

**After this issue is resolved, consider:**

1. **Visualization** - Add charts showing cluster assignments, PCA projections
2. **Parameter Tuning** - Allow users to adjust k, learning rate, n_components
3. **Model Comparison** - Side-by-side results for different algorithms
4. **Export Results** - Download predictions/transformations as CSV
5. **Streaming Data** - Handle large CSVs in chunks to avoid memory issues

**Extensibility:**
The same pattern can be used to add new algorithms:
1. Implement algorithm in separate crate with trait
2. Add to Algorithm enum
3. Create run_[algorithm]() function following the pattern
4. Add to dispatcher

---

**Estimated Effort:** 1-2 hours (mostly testing with different CSV files)

**Files Modified:** 1 (\`web/src/components/ml_playground.rs\`)

**Lines Changed:** ~30 (remove comments, simplify 10 method calls, update dispatcher)
